File: samples/SelfAspNetVb/SelfAspNetVb/Chap09/SqlSiteMapProvider.vb
Imports System.Data.SqlClient

Public Class SqlSiteMapProvider : Inherits StaticSiteMapProvider
    Private connectName As String = Nothing
    Private root As SiteMapNode = Nothing

    Public Overrides Sub Initialize(name As String, attributes As NameValueCollection)
        If connectName Is Nothing Then
            MyBase.Initialize(name, attributes)
            connectName = attributes("connectName")
        End If
    End Sub

    Public Overrides Function BuildSiteMap() As System.Web.SiteMapNode
        If root Is Nothing Then
            SyncLock Me
                Me.CreateNode(root)
            End SyncLock
        End If
        Return root
    End Function

    Protected Overrides Function GetRootNodeCore() As System.Web.SiteMapNode
        Return Me.BuildSiteMap()
    End Function

    Private Sub CreateNode(pnode As SiteMapNode)
        Dim setting As ConnectionStringSettings =
          ConfigurationManager.ConnectionStrings(Me.connectName)
        Using db As New SqlConnection(setting.ConnectionString)
            Dim comm As New SqlCommand("SELECT mid, url, title, description FROM Sitemap WHERE parent = @parent", db)
            comm.Parameters.AddWithValue("@parent", If(pnode Is Nothing, 0, pnode.Key))
            db.Open()
            Dim reader As SqlDataReader = comm.ExecuteReader()
            Do While reader.Read()
                If pnode Is Nothing Then
                    root = New SiteMapNode(Me, reader("mid"), reader("url"), reader("title"), reader("description"))
                    Me.AddNode(root)
                    Me.CreateNode(root)
                Else
                    Dim node As New SiteMapNode(Me, reader("mid"), reader("url"), reader("title"), reader("description"))
                    Me.AddNode(node, pnode)
                    Me.CreateNode(node)
                End If
            Loop
        End Using
    End Sub
End Class

File: samples/SelfAspNet/SelfAspNet/Chap09/SqlSiteMapProvider.cs
using System;
using System.Collections.Specialized;
using System.Configuration;
using System.Data.SqlClient;
using System.Web;

namespace SelfAspNet.Chap09
{
    public class SqlSiteMapProvider : StaticSiteMapProvider
    {
        private String connectName;
        private SiteMapNode root;

        public override void Initialize(string name, NameValueCollection attributes)
        {
            if (connectName == null)
            {
                base.Initialize(name, attributes);
                connectName = attributes["connectName"];
            }
        }

        public override SiteMapNode BuildSiteMap()
        {
            if (root == null)
            {
                lock (this)
                {
                    this.CreateNode(root);
                }
            }
            return root;
        }

        protected override SiteMapNode GetRootNodeCore()
        {
            return this.BuildSiteMap();
        }

        private void CreateNode(SiteMapNode pnode)
        {
            ConnectionStringSettings setting =
                ConfigurationManager.ConnectionStrings[this.connectName];
            using (var db = new SqlConnection(setting.ConnectionString))
            {
                var comm = new SqlCommand("SELECT mid, url, title, description FROM Sitemap WHERE parent = @parent", db);
                comm.Parameters.AddWithValue("@parent", pnode == null ? 0 : Int32.Parse(pnode.Key));
                db.Open();
                var reader = comm.ExecuteReader();
                while (reader.Read())
                {
                    if (pnode == null)
                    {
                        root = new SiteMapNode(this, reader["mid"].ToString(), reader["url"].ToString(), reader["title"].ToString(), reader["description"].ToString());
                        this.AddNode(root);
                        this.CreateNode(root);
                    }
                    else
                    {
                        var node = new SiteMapNode(this, reader["mid"].ToString(), reader["url"].ToString(), reader["title"].ToString(), reader["description"].ToString());
                        this.AddNode(node, pnode);
                        this.CreateNode(node);
                    }
                }
            }
        }
    }
}